### ISSUE TITLE
wesnothd: reconnection after hard DC leaves dirty state

### ISSUE THREAD

**Issue by vgaming** [wesnoth/wesnoth#4011]
### Steps:
* start game as player A
* wait for other people to join
* disconnect the hard way, e.g. force-kill the app `kill -9 process_id`
* join back

### Observed behavior:
* For users that joined your game, you'll remain in game. Even though your re-connected and might observe a different game by that moment already.
* You can get into your original game by trying to observe it. If you get out of such game, the other joined players will still see you as active.

Basically what I think is happening, is that DC + reconnection does not lead to the usual cleanup procedures as if user disconnected normally.

The issue feels quite nasty to me as the only thing needed for that to happen is a DC by one player.

**Comment by loonycyborg** [comment]
There are probably some other conditions to reproduce this. If I just terminate starting client with SIGKILL at game creation state it is aborted immediately even if there are other players who have joined.

**Comment by vgaming** [comment]
@loonycyborg interesting. It certainly works for me (ArchLinux amd64). Can you maybe try with a classical DC, e.g. disconnect your internet for a second?

I can help testing if you want, either DC myself or let you do it. (Write me on IRC.)

**Comment by loonycyborg** [comment]
I saw this bug in action, will look into it more.

**Comment by loonycyborg** [comment]
I've rebuilt "trunk" server, that is the one the client built from master branch connects to by default, so it contains this fix. Some testing by more people for fixing of this bug/side effects of fix would be cool.

**Comment by vgaming** [comment]
@loonycyborg thanks! I'll try to test in a few hours or tomorrow.

**Comment by vgaming** [comment]
I've tested it with a freshly built wesnoth (master branch). The DC bug cannot be reproduced by me anymore!

Thanks a lot for the fix!
